src/infra/core/tests/tests/ingest/test_fetch.rs

```diff
@@ -686,6 +630,387 @@ async fn test_fetch_container_ok() {
+    let temp_dir = tempfile::tempdir().unwrap();
+    let target_path = temp_dir.path().join("fetched.bin");
+
+    let fetch_svc = FetchService::new(
+        Arc::new(ContainerRuntime::default()),
+        temp_dir.path().join("run"),
+    );
+    let custom_batch_size = 40;
+
+    let prev_source_state = {
+        let fetch_step_1 = FetchStep::Container(FetchStepContainer {
+            image: BUSYBOX.to_owned(),
+            command: Some(vec!["sh".to_owned()]),
+            args: Some(vec!["-c".to_owned(), HAS_MORE_TESTER_SCRIPT.to_owned()]),
+            env: Some(vec![EnvVar {
+                name: ODF_BATCH_SIZE.to_owned(),
+                value: Some(custom_batch_size.to_string()),
+            }]),
+        });
+
+        let res_1 = fetch_svc
+            .fetch("1", &fetch_step_1, None, &target_path, &Utc::now(), None)
+            .await
+            .unwrap();
+
+        assert!(target_path.exists());
+        assert_matches!(
+            &res_1,
+            FetchResult::Updated(FetchResultUpdated {
+                source_state: Some(PollingSourceState::ETag(expected_etag)),
+                has_more: true,
+                ..
+            }) if expected_etag == "40"
+        );
+        assert!(target_path.exists());
+        assert_eq!(
+            std::fs::read_to_string(&target_path).unwrap(),
+            CSV_BATCH_OUTPUT
+        );
+
+        match res_1 {
+            FetchResult::Updated(x) => x.source_state,
+            _ => unreachable!(),
+        }
+    };
+    let prev_source_state = {
+        let fetch_step_2 = FetchStep::Container(FetchStepContainer {
+            image: BUSYBOX.to_owned(),
+            command: Some(vec!["sh".to_owned()]),
+            args: Some(vec!["-c".to_owned(), HAS_MORE_TESTER_SCRIPT.to_owned()]),
+            env: Some(vec![EnvVar {
+                name: ODF_BATCH_SIZE.to_owned(),
+                value: Some(custom_batch_size.to_string()),
+            }]),
+        });
+
+        let res_2 = fetch_svc
+            .fetch(
+                "2",
+                &fetch_step_2,
+                prev_source_state.as_ref(),
+                &target_path,
+                &Utc::now(),
+                None,
+            )
+            .await
+            .unwrap();
+
+        assert_matches!(
+            &res_2,
+            FetchResult::Updated(FetchResultUpdated {
+                source_state: Some(PollingSourceState::ETag(expected_etag)),
+                has_more: true,
+                ..
+            }) if expected_etag == "80"
+        );
+        assert!(target_path.exists());
+        assert_eq!(
+            std::fs::read_to_string(&target_path).unwrap(),
+            CSV_BATCH_OUTPUT
+        );
+
+        match res_2 {
+            FetchResult::Updated(x) => x.source_state,
+            _ => unreachable!(),
+        }
+    };
+    let prev_source_state = {
+        let fetch_step_3 = FetchStep::Container(FetchStepContainer {
```

Comment:
To me it seems like two iterations (one expecting `has_more==true`  another `false`) is enough here. If 3rd and 4th iterations are somehow special please put a comment above them explaining what is being checked.

Reply:
Added comments to each step, I hope it looks like more clear -- **Q:** What do you think?

Reply:
I think 2nd iteration doesn't test any new boundary conditions that are not covered elsewhere and could be removed to make the test shorter, but I'm fine with it either way.